src/libraries/System.Net.Quic/src/System/Net/Quic/QuicStream.cs

```diff
@@ -360,19 +361,34 @@ public ValueTask WriteAsync(ReadOnlyMemory<byte> buffer, bool completeWrites, Ca
             return valueTask;
         }
 
-        _sendBuffers.Initialize(buffer);
-        unsafe
+        lock (_sendBuffersLock)
         {
-            int status = MsQuicApi.Api.ApiTable->StreamSend(
-                _handle.QuicHandle,
-                _sendBuffers.Buffers,
-                (uint)_sendBuffers.Count,
-                completeWrites ? QUIC_SEND_FLAGS.FIN : QUIC_SEND_FLAGS.NONE,
-                null);
-            if (ThrowHelper.TryGetStreamExceptionForMsQuicStatus(status, out Exception? exception))
+            ObjectDisposedException.ThrowIf(_disposed == 1, this); // TODO: valueTask is left unobserved
+            unsafe
             {
-                _sendBuffers.Reset();
-                _sendTcs.TrySetException(exception, final: true);
+                if (_sendBuffers.Count > 0 && _sendBuffers.Buffers[0].Buffer != null)
```

Comment:
I believe this will become a non-issue when https://github.com/dotnet/runtime/issues/73691 is implemented. I made it a high prio issue for 8.0 since it seems more important now.